pypsa/inverter_functions.py

```diff
@@ -0,0 +1,512 @@
+"""importing important libraries."""
+import logging
+import numpy as np
+from copy import deepcopy
+logger = logging.getLogger(__name__)
+
+
+def fixed_cosphi(p_input, now, c_attrs, c_list_name):
+    """
+    fix power factor inverter controller.
+    reference : https://www.academia.edu/24772355/
+
+    Parameters
+    ----------
+    p_input : pandas data frame
+        Input active power for controller, comming from n.component_t.p.
+    now : single snapshot
+        Current  element of n.snapshots on which the power flow is run.
+    c_attrs : pandas data frame
+        Component attrs including controller required parameters for controlled
+        indexes, i.e. power_factor choice for generators, loads...
+    c_list_name : string
+        Component name, i.e. 'loads', 'storage_units', 'generators'.
+
+    Returns
+    -------
+    q_set : pandas data frame
+        Is the new reactive power that will be set as new q_set in each
+        controlled component as a result of applying this controller.
+    """
+    # needed parameters
+    p = p_input.loc[now, c_attrs.index]
+    power_factor = c_attrs['power_factor']
+    s_nom = c_attrs['s_nom']
+    setting_p_set_required = False
+    new_p_set = 0
+    Q_inv_cap, Q_allowable, Q_max = find_allowable_q(p, power_factor, s_nom)
+    q_set = -Q_allowable
+
+    # check if the calculated Q_max is not exceeding the inverter capacity \ if
+    # yes then increase it based on inverter capacity and reset p_set.
+    if (abs(Q_max) > Q_inv_cap).any().any():
+        setting_p_set_required = True
+        new_p_set = adjust_p_set(s_nom, q_set, p, c_list_name, 'fixed_cosphi')
+
+    return q_set, new_p_set, setting_p_set_required
+
+
+def cosphi_p(p_input, now, df, df_t, c_attrs, time_varying_p_set, c_list_name):
+    """
+    Power factor as a function of active power (cosphi_p) controller.
+    reference : https://ieeexplore.ieee.org/document/6096349.
```

Comment:
Same as above. Provide a brief description of the control strategy here, but feel free to keep the paper as additional reference. Maybe provide the doi link instead.